Iter13

MAGeorg opened the pull request.

Commit:
iter10: added ping endpoint

Commit:
added makefile and local file for golangci

Commit:
refactor code

Commit:
iter11: added simple migrator

Commit:
iter11: added work with DB

1. added work with postgres
2. fixed migration
3. fixed DB scheme

Commit:
iter11: fixed error

Commit:
added docker-compose for local test

Commit:
fixed var

Commit:
added logging

Commit:
add custom golangci-lint file

1. added file for linter
2. fixed error from linter

Commit:
iter13: added functional

1. added contect to handlers
2. added skeleton new endpoint
3. fixed test

Commit:
iter13: added banch endpoint

Commit:
iter13: fixed schema db

Commit:
iter13: added go-migration

Commit:
iter13: added check unique

arttet commented on .golangci.yml:
```diff
@@ -0,0 +1,252 @@
+  govet:
+    enable-all: true
+    disable:
+      - fieldalignment # too strict
```
NICE TO HAVE: I would enable it.

arttet commented on .golangci.yml:
```diff
@@ -0,0 +1,252 @@
+run:
```
Good job!

arttet commented on cmd/shortener/main.go:
```diff
@@ -1,11 +1,22 @@
+
+const (
+	// константа для считывания миграций, если используется чистый sql
+	source = "../../internal/storage/migration/postgres/001.init_schema.sql"
```
CHANGE REQUEST: this is a magic string.

arttet commented on internal/logger/logger.go:
```diff
@@ -6,8 +7,10 @@ import (
 )
 
+// Sugar логгер
 var Sugar zap.SugaredLogger
```
CHANGE REQUEST: please don't use global variables

MAGeorg replied:
это не используемая переменная осталась где-то с iter9, забыл убрать)

arttet commented on internal/storage/db.go:
```diff
@@ -0,0 +1,86 @@
+		"INSERT INTO shot_url (hash_value, origin_url) VALUES ($1,$2);", h, url)
+
+	// проверка на дубликат
+	if driverErr, ok := err.(*pgconn.PgError); ok && driverErr.Code == "23505" {
```
NICE TO HAVE: it would be better if you try to use `INSERT .. ON CONFLICT`

arttet commented on internal/storage/migration/migration.go:
```diff
@@ -0,0 +1,111 @@
+}
+
+// проверка, что схемы еще нет
+func CheckExistScheme(ctx context.Context, db *sql.DB) bool {
```
CHANGE REQUEST: a migrator should do it

arttet commented on internal/storage/migration/postgres/001.init_schema.sql:
```diff
@@ -0,0 +1,11 @@
+CREATE TABLE shot_url (
+    id         SERIAL  NOT NULL,
+    hash_value BIGINT NOT NULL,
+    origin_url TEXT    UNIQUE,
```
```suggestion
    origin_url VARCHAR(2048) UNIQUE,
```

arttet commented on internal/storage/migration/postgres/001.init_schema.sql:
```diff
@@ -0,0 +1,11 @@
+
+CREATE TABLE shot_url (
+    id         SERIAL  NOT NULL,
+    hash_value BIGINT NOT NULL,
```
```suggestion
    hash_value smallserial NOT NULL,
```

MAGeorg replied:
smallserial is too small to store uint32, perhaps you mean bigserial, bigserial does not use negative numbers, it will be more beneficial to use than bigint

MAGeorg replied:
> 
smallserial is too small to store uint32, perhaps you mean bigserial, bigserial does not use negative numbers, it will be more beneficial to use than bigint

arttet commented on internal/storage/migration/postgres/init_schema.go:
```diff
@@ -0,0 +1,21 @@
+
+// функция Up возвращает sql строку для создания схемы
+func Up() string {
+	return `
```
CHANGE REQUEST: please DRY

arttet reviewed (approved):
Please fix `CodeCov` and `dependabot`

```
Error: Codecov token not found. Please provide Codecov token with -t flag.
```

Commit:
iter13: fixed some comment

1. changed scheme db
2. removed unusable global var logger
3. uncomment some linters
4. fixed error from linters
5. check exist scheme moved from main to migrator
